ipa-core/src/protocol/dp/distributions.rs

```diff
@@ -63,44 +58,92 @@ impl Distribution<f64> for RoundedBoxMuller {
         self.inner.sample(rng).round()
     }
 }
-
 impl From<BoxMuller> for RoundedBoxMuller {
     fn from(value: BoxMuller) -> Self {
         Self { inner: value }
     }
 }
-
+/// Double Geometric Distribution
+///
+/// Generates a sample from a geometric distribution with the given success probability.
+fn generate_geometric<R: Rng + ?Sized>(probability: f64, rng: &mut R) -> isize {
+    // Create a Bernoulli distribution with the specified success probability
+    let bernoulli = Bernoulli::new(probability).expect("Invalid probability");
+    // Generate Bernoulli random numbers until the first success
+    // let mut rng = rand::thread_rng();
+    let mut attempts = 0;
+    // let sample = ud.sample(rng);
+    while !bernoulli.sample(rng) {
+        attempts += 1;
+    }
+    attempts
+}
+/// Generates a sample from a double geometric distribution with the given success probability and shift parameter.
+fn generate_double_geometric<R: Rng + ?Sized>(s: f64, shift: isize, rng: &mut R) -> isize {
+    let success_probability = 1.0 - E.powf(-1.0 / s);
+    let attempts1 = generate_geometric(success_probability, rng);
+    let attempts2 = generate_geometric(success_probability, rng);
+    shift + attempts1 - attempts2
+}
+/// Generates a sample from a double geometric distribution with the given success probability and shift parameter.
+fn generate_truncated_double_geometric<R: Rng + ?Sized>(s: f64, n: isize, rng: &mut R) -> isize {
+    let mut reject = 1;
+    let mut sample = 0; // Declare sample here
+    while reject == 1 {
+        sample = generate_double_geometric(s, n, rng); // Assign a value to sample inside the loop
+        if sample >= 0 && sample <= (2 * n) {
+            reject = 0;
+        }
+    }
+    sample // Return the final value of sample
+}
+/// Truncated Double Geometric distribution.
+#[derive(Debug)]
+pub struct TruncatedDoubleGeometric {
+    s: f64,
+    shift: isize,
+}
+impl TruncatedDoubleGeometric {
+    /// Creates a new `TruncatedDoubleGeometric` distribution with the given success probability and shift parameter.
+    pub fn new(s: f64, shift: isize) -> Self {
```

Comment:
likewise, it is better to reject invalid values during the construction phase, rather than panicking at `sample` call

Reply:
updated to return Result type